source/Halibut/Transport/Protocol/MessageExchangeStream.cs

```diff
@@ -355,4 +411,107 @@ void SetShortTimeouts()
             stream.ReadTimeout = (int)HalibutLimits.TcpClientHeartbeatReceiveTimeout.TotalMilliseconds;
         }
     }
+
+    class BufferedReadStream : Stream
```

Comment:
We should extract to a different file

Reply:
I've pulled these out into their own files, renamed them from "BufferXStream" to "CaptureXStream" to better reflect what they do, and added some tests.